File: app/src/main/java/at/droelf/travellogapp/NetworkClient.java
package at.droelf.travellogapp;

import android.util.Log;

import org.joda.time.DateTimeZone;
import org.joda.time.LocalDateTime;
import org.springframework.core.io.FileSystemResource;
import org.springframework.http.HttpAuthentication;
import org.springframework.http.HttpBasicAuthentication;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.FormHttpMessageConverter;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.http.converter.ResourceHttpMessageConverter;
import org.springframework.http.converter.StringHttpMessageConverter;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;

import java.util.List;

public class NetworkClient {

    private final RestTemplate restTemplate;

    private final String apiImageUploadBase = "http://k2:9000/api/images/uploadImage";
    private final String imageFileFormKey = "file";

    private final String apiGpxDataUpload = "http://k2:9000/api/tracks/upload/";

    private final String username = "admin";
    private final String password = "1234";

    private final HttpHeaders requestHeadersWithAuth;

    public NetworkClient() {
        final HttpAuthentication authHeader = new HttpBasicAuthentication(username, password);

        requestHeadersWithAuth = new HttpHeaders();
        requestHeadersWithAuth.setAuthorization(authHeader);


        restTemplate = new RestTemplate();

        final List<HttpMessageConverter<?>> messageConverters = restTemplate.getMessageConverters();
        messageConverters.add(new ResourceHttpMessageConverter());
        messageConverters.add(new StringHttpMessageConverter());
        messageConverters.add(new FormHttpMessageConverter());
    }

    boolean uploadImage(LocalDateTime dateTime, String timezone, String name, String imagePath) {
        final String url = apiImageUploadBase + "/" + getLocalDateTimeAsStringForImageUpload(dateTime) + timezone + "/" + name;

        return putFileToUrlWithAuth(url, imagePath);
    }

    boolean uploadGpxData(String activity, String timezone, String pathToFile) {
        final String url = apiGpxDataUpload + "/" + timezone + "/" + activity;

        return putFileToUrlWithAuth(url, pathToFile);
    }

    boolean putFileToUrlWithAuth(String url, String filePath) {
        final MultiValueMap<String, Object> parts = new LinkedMultiValueMap<String, Object>();
        parts.add(imageFileFormKey, new FileSystemResource(filePath));

        final HttpEntity<?> requestEntity = new HttpEntity<Object>(parts, requestHeadersWithAuth);

        Log.d("Network", "Putting file " + filePath + " to " + url);

        ResponseEntity<Void> response = restTemplate.exchange(url, HttpMethod.PUT, requestEntity, Void.class);
        return response.getStatusCode() == HttpStatus.OK;
    }

    private String getTimeZoneForGpxDataUpload(DateTimeZone dateTimeZone) {
        return "";
    }

    private String getLocalDateTimeAsStringForImageUpload(LocalDateTime dateTime) {
        return DateTimeUtils.localDateTimeToIsoString(dateTime);
    }
}

File: app/src/main/scala/at/droelf/travellogapp/NetworkClient.scala
package at.droelf.travellogapp

import android.util.Log
import org.joda.time.{DateTimeZone, LocalDateTime}
import org.springframework.core.io.FileSystemResource
import org.springframework.http.converter.{HttpMessageConverter, FormHttpMessageConverter, StringHttpMessageConverter, ResourceHttpMessageConverter}
import org.springframework.http._
import org.springframework.util.{LinkedMultiValueMap, MultiValueMap}
import org.springframework.web.client.RestTemplate

class NetworkClient {

    val authHeader: HttpAuthentication = new HttpBasicAuthentication(username, password)
    val requestHeadersWithAuth = new HttpHeaders
    requestHeadersWithAuth.setAuthorization(authHeader)
    val restTemplate = new RestTemplate
    val messageConverters: java.util.List[HttpMessageConverter[_]] = restTemplate.getMessageConverters
    messageConverters.add(new ResourceHttpMessageConverter)
    messageConverters.add(new StringHttpMessageConverter)
    messageConverters.add(new FormHttpMessageConverter)

  private[travellogapp] def uploadImage(dateTime: LocalDateTime, timezone: String, name: String, imagePath: String): Boolean = {
    val url: String = apiImageUploadBase + "/" + getLocalDateTimeAsStringForImageUpload(dateTime) + timezone + "/" + name
    return putFileToUrlWithAuth(url, imagePath)
  }

  private[travellogapp] def uploadGpxData(activity: String, timezone: String, pathToFile: String): Boolean = {
    val url: String = apiGpxDataUpload + "/" + timezone + "/" + activity
    return putFileToUrlWithAuth(url, pathToFile)
  }

  private[travellogapp] def putFileToUrlWithAuth(url: String, filePath: String): Boolean = {
    val parts: MultiValueMap[String, AnyRef] = new LinkedMultiValueMap[String, AnyRef]
    parts.add(imageFileFormKey, new FileSystemResource(filePath))
    val requestEntity: HttpEntity[_] = new HttpEntity[AnyRef](parts, requestHeadersWithAuth)
    Log.d("Network", "Putting file " + filePath + " to " + url)
    val response: ResponseEntity[Void] = restTemplate.exchange(url, HttpMethod.PUT, requestEntity, classOf[Void])
    return response.getStatusCode eq HttpStatus.OK
  }

  private def getTimeZoneForGpxDataUpload(dateTimeZone: DateTimeZone): String = {
    return ""
  }

  private def getLocalDateTimeAsStringForImageUpload(dateTime: LocalDateTime): String = {
    return DateTimeUtils.localDateTimeToIsoString(dateTime)
  }

  private final val apiImageUploadBase: String = "http://i8:9000/api/images/uploadImage"
  private final val imageFileFormKey: String = "file"
  private final val apiGpxDataUpload: String = "http://i8:9000/api/tracks/upload/"
  private final val username: String = "admin"
  private final val password: String = "1234"
}